agent/consul/acl_endpoint.go

```diff
@@ -2,239 +2,866 @@ package consul
 
 import (
 	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
 	"time"
 
 	"github.com/armon/go-metrics"
 	"github.com/hashicorp/consul/acl"
 	"github.com/hashicorp/consul/agent/consul/state"
 	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/lib"
 	"github.com/hashicorp/go-memdb"
 	"github.com/hashicorp/go-uuid"
 )
 
+const (
+	// aclBootstrapReset is the file name to create in the data dir. It's only contents
+	// should be the reset index
+	aclBootstrapReset = "acl-bootstrap-reset"
+)
+
+// Regex for matching
+var validPolicyName = regexp.MustCompile(`^[A-Za-z0-9\-_]{1,128}$`)
+
 // ACL endpoint is used to manipulate ACLs
 type ACL struct {
 	srv *Server
 }
 
+func (a *ACL) fileBootstrapResetIndex() uint64 {
+	// Determine the file path to check
+	path := filepath.Join(a.srv.config.DataDir, aclBootstrapReset)
+
+	// Read the file
+	raw, err := ioutil.ReadFile(path)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			a.srv.logger.Printf("[ERR] acl.bootstrap: failed to read %q: %v", path, err)
+		}
+		return 0
+	}
+
+	// Attempt to parse the file
+	var resetIdx uint64
+	if _, err := fmt.Sscanf(string(raw), "%d", &resetIdx); err != nil {
+		a.srv.logger.Printf("[ERR] acl.bootstrap: failed to parse %q: %v", path, err)
+		return 0
+	}
+
+	// Return the reset index
+	a.srv.logger.Printf("[WARN] acl.bootstrap: parsed %q: reset index %d", path, resetIdx)
+	return resetIdx
```

Comment:
It doesn't really matter what the index is. The general bootstrap reset procedure is to make a bootstrap request which will tell you the cluster already has bootstrapped ACLs and tells you what the reset index is. You then write that index into the reset index file and rerun bootstrapping. This time it lets it work because the index in the file on disk matches whats in the index table.

One thing I will add is to remove that file whenever bootstrapping is successful to prevent any oddity if a old snapshot is restored.

Reply:
👍 I didn't figure out that this was a user-written file used as an escape hatch to re-bootstrap. That's fine, I was assuming in these comments it was somehow used/written to by Consul in some circumstance.